kernel/objects/capability-spinning/objects/CapEntry.hh

```diff
@@ -83,22 +91,81 @@ namespace mythos {
      * allocated. Returns true if zombified. */ 
     bool kill();
 
-    optional<void> unlink();
-    bool try_lock() { return !(_next.fetch_or(LOCKED_FLAG) & LOCKED_FLAG); }
-    void lock() { while (!try_lock()) { hwthread_pause(); } }
-    void unlock() { auto res = _next.fetch_and(~LOCKED_FLAG); ASSERT(res & LOCKED_FLAG); }
+    bool kill(Cap expected);
+
+    optional<void> unlinkAndUnlockLinks();
+
+    /* lock next functions protect the link to the next CapEntry */
+
+    bool try_lock_next()
+    { 
+      bool ret = !(_next.fetch_or(LOCKED_FLAG) & LOCKED_FLAG);
+      MLOG_ERROR(mlog::cap, __PRETTY_FUNCTION__, DVAR(this), ret? " locked" : "locking failed!");
+      return ret;
+    }
+
+    void lock_next()
```

Comment:
`kill()` does not call `lock_next()`, you probable mean the `RevokeOperation`?

> We retry to acquire the lock for unbounded time, because at some point in time no entry can be left behind is for our predecessor to be killed. 

I don't understand what you mean with this sentence. Every operation that `lock_next` unlocks it a short time after.


`DELETED_FLAG` means that the `CapEntry`(!) is soon going to be deleted (as part of a soon-to-be-deleted object), it should not be set for a `CapEntry` that is going to be reused. Killing an entry with (`kill(), try_kill(CapEntry*)`) is enough to ensure nobody inherit from that cap. Therefore never will insert after this entry until a new cap is in there.

Reply:
yes, I used `kill` for the wrong thing. Thought that kill does the whole "remove entry from chain" business and is called after the entry's capability and kernel object is prepared for deletion. Have to read more of the code obviously :-)

Reply:
kill ensures that no new child can be created and inserted after us. Excellent. What about concurrent inherit operations: If I called kill() on a leaf entry, is it necessary to check whether someone concurrently inserted a child that is now the new leaf? 

My concern is, when a revoke is aborted due to concurrency. It leaves many capability entries in a killed state where nothing except retrying the revoke can be done with them. I am fine with that. It is not counterintuitive, but some users might not expect that an aborted revoke leaves the capabilities in a "broken" state.